### ISSUE TITLE
mergeObjects has broken type definition

### ISSUE THREAD

**Issue by jam-fran** [seasonedcc/domain-functions#81]
It appears that something's broken with `mergeObjects`' type definition in the generated `utils.d.ts` file. I've pasted portion of it below (only included a tiny sample, in the actual file it's thousands and thousands of lines long). In my IDE importing the function crashes the Typescript server and breaks all inferred types. 

`declare function mergeObjects<T extends unknown[] = unknown[]>(objs: T): (T extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_10]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_10 ? { [K_9 in keyof T_10]: (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_10]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_9]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_9 ? { [K_8 in keyof T_9]: (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_10]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_10 ? { [K_9 in keyof T_10]: (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_10]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_9]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_8]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_8 ? { [K_7 in keyof T_8]: (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_10]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_10 ? { [K_9 in keyof T_10]: (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_10]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_9]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_9 ? { [K_8 in keyof T_9]: (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)[K_10]; } : never : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_10 ? { [K_9 in keyof T_10]: (rest extends [infer first, ...infer rest] ? (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first) extends infer T_11 ? { [K_10 in keyof T_11]: (rest extends [infer first, ...infer rest] ? any : Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<Omit<{}, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first, keyof first> & first)`

**Comment by gustavoguichard** [comment]
Hey @jam-fran , thanks for reporting it. I'm gonna take a look at it later this week.

**Comment by gustavoguichard** [comment]
Hey @jam-fran , I investigated a little bit and realized the source of this bug is the `Prettify` type which should make the result of `mergeObjects` more readable. We'll probably have to rollback on that and sacrifice readability.

Remix team has faced that recently and they also had to [rollback](https://github.com/remix-run/remix/pull/6736), check out [the discussion](https://github.com/remix-run/remix/pull/6746#issuecomment-1619068201) if you want to learn more about it.

Thanks again for reporting it and I'll soon send the PR rolling it back.

**Comment by gustavoguichard** [comment]
Hey @jam-fran , it looks like the change will be a lil harder than I thought, could you provide me with an example that breaks the type so I can add a proper regression test for that?

**Comment by jam-fran** [comment]
Hi @gustavoguichard -- thanks for looking into this. 

Here's a minimal reproduction:

```
const funcA = makeDomainFunction(
  z.object({
    argA: z.boolean(),
    argB: z.string(),
  }),
)(({ argA, argB }) => {
  return { argA, argB }
})

const funcB = makeDomainFunction(
  z.object({
    argA: z.boolean(),
    argB: z.string(),
  }),
)(({ argA, argB }) => {
  return { foo: 'bar' }
})

const funcC = map(sequence(funcA, funcB), mergeObjects)
```

Here, typescript mistakenly thinks the result type of funcC is:

```
const funcC: DomainFunction<{
    foo: string;
}>
```

Hope this helps -- let me know if you need anything else.

**Comment by gustavoguichard** [comment]
Hi @jam-fran , thanks for clarifying! It helped a lot!

I strongly believe I fixed the bug in domain-functions@1.8.0 , reopen the issue and LMK if that's not the case though ;)

The file with that problem used to be this one: https://www.unpkg.com/domain-functions@1.7.0/types/utils.d.ts
In the new version we can see how it looks a lot better already: https://www.unpkg.com/domain-functions@1.8.0/types/utils.d.ts

You can also see a reproduction of your code above in the [TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbziAhgawKYBEKuAOwDEBXfAYxmAnwBo4BnDAR2I3IztTE4ygHMMAeQBGAKwwV6cAL5wAZlFxwARABNcKAgFo5pClXz1lAbgBQoSLERwAXjPmKQcAOQ2Iq52dNlq9eLvIAQTgAXmR0bA0CEnJKagAKUzhbADoIMQkYeIQk5LgUfkCALlThCAgAGwwUfHiAShpc5IK+ACESmxS-KAI+esbk6QbTOvjs-MK6FtaZOtCAPkRcqAwYYih8axbAqf4Z6VMh029ffz0ZsNRMHDwiPTja3M708Qpspom+YtLyqpr+j7TDpdGA9fB9YaDYajcbbXZtWYLJbJFZrDbWOTlErOYQFZwyQ51Y4+QxncgAYVC4TA8UYLDYZAw8QCZB28nODWQvAEIleMHoRKAA). It seems to be working now 🚀

**Comment by jam-fran** [comment]
Looks great, thanks so much for your responsiveness here @gustavoguichard
